src/mono/mono/mini/aot-compiler.c

```diff
@@ -6128,6 +6130,63 @@ method_is_externally_callable (MonoAotCompile *acfg, MonoMethod *method)
 	}
 }
 
+#ifdef MONO_ARCH_AOT_SUPPORTED
+static const char *
+get_pinvoke_import (MonoAotCompile *acfg, MonoMethod *method)
+{
+	MonoImage *image = m_class_get_image (method->klass);
+	MonoMethodPInvoke *piinfo = (MonoMethodPInvoke *) method;
+	MonoTableInfo *tables = image->tables;
+	MonoTableInfo *im = &tables [MONO_TABLE_IMPLMAP];
+	guint32 im_cols [MONO_IMPLMAP_SIZE];
+	char *import;
+
+	import = (char *)g_hash_table_lookup (acfg->method_to_pinvoke_import, method);
+	if (import != NULL)
+		return import;
+
+	if (piinfo->implmap_idx == 0 || mono_metadata_table_bounds_check (image, MONO_TABLE_IMPLMAP, piinfo->implmap_idx))
+		return NULL;
+
+	mono_metadata_decode_row (im, piinfo->implmap_idx - 1, im_cols, MONO_IMPLMAP_SIZE);
+
+	int module_idx = im_cols [MONO_IMPLMAP_SCOPE];
+	if (module_idx == 0 || mono_metadata_table_bounds_check (image, MONO_TABLE_MODULEREF, module_idx))
+		return NULL;
+
+	import = g_strdup_printf ("%s", mono_metadata_string_heap (image, im_cols [MONO_IMPLMAP_NAME]));
+
+	g_hash_table_insert (acfg->method_to_pinvoke_import, method, import);
+
+	return import;
+}
+#else
+static const char *
+get_pinvoke_import (MonoAotCompile *acfg, MonoMethod *method)
+{
+	return NULL;
+}
+#endif
+
+/*
+ * mono_aot_direct_pinvoke_enabled_for_method
+ *
+ * Return whether the method is specified to be directly pinvoked
+ */
+static gboolean
+mono_aot_direct_pinvoke_enabled_for_method (MonoAotCompile *acfg, MonoMethod *method)
+{
+	const char *sym = get_pinvoke_import (acfg, method);
+	const char *module_name = acfg->image->module_name;
+	if (g_hash_table_contains (acfg->direct_pinvokes, module_name)) {
+		GHashTable *val = g_hash_table_lookup (acfg->direct_pinvokes, module_name);
+		if (!val)
+			return TRUE;
```

Comment:
I guess this should return FALSE? If it doesn't find the module, then its not a direct pinvoked method.

Reply:
If we reach `if (!val)`, we should have a `key:value` pair of `module:NULL` in the `acfg->direct_pinvokes` HashTable. This would lead to all `entrypoints` of `module` being direct callable.

Reply:
So is this the case where user have used a module entry meaning all entrypoints shoud be direct pinvoked in the module? I didn't understand that a module entry in hash table with null value represented all entrypoints in the module. Maybe add a comment to clarify that design a little? You can also combine the contains and lookup using g_hash_table_lookup_extended .

Reply:
Yes, this is where a specified direct pinvoke `<module>` would trump any specified direct pinvoke `<module>!<entrypoint>`. 

The intention is if a provided direct pinvoke does not have the `!` delimiter, it is considered a `<module>` and is inserted into the hash table as Key:Value `<module>:NULL`. If later down the line a `<module>!<entrypoint>` is provided, instead of recording the given `<entrypoint>` (by adding it to the "set" EntryPoint HashTable), it will early return.

This follows the NativeAOT approach https://github.com/dotnet/runtime/blob/ab2d195f169a08bf8eb978b472b2d3aca61b0a46/src/coreclr/tools/aot/ILCompiler/ConfigurablePInvokePolicy.cs#L64-L65

I'll add comments/descriptions of the functions in soon, didnt' have the time to get around to it yet and I wanted to just get the bulk logic out quickly.